README.md
# SpectroGraphic
_Turn any image into a sound whose spectrogram looks like the image!_

![result](banner.png)

<hr>

#### What is this?

Most sounds are intricate combinations of many acoustic waves each having different frequencies and intensities. A spectrogram is a way to represent sound by plotting time on the horizontal axis and the frequency spectrum on the vertical axis. Sort of like sheet music on steroids.

What this tool does is, taking an image and simply interpreting it as a spectrogram. Therefore, by generating the corresponding sound, we have embedded our image in a spectrogram.

The game DOOM used a similar technique to [hide satanic figures inside its soundtrack](https://www.theverge.com/2016/5/31/11825606/doom-2016-soundtrack-satan-666-inverted-pentagram). Now everyone can do the same! 😊

#### Set-up

Get the command-line tool `spectrographic` via `pip` by running `pip install spectrographic`. You can also simply use `spectrographic.py` from `stand-alone\` as a command-line tool directly.
Furthermore, make sure you meet all the dependencies inside the `requirements.txt`. Install them with `pip install requirements.txt`.

After installation with `pip` one simply needs to run `spectrographic [...]` in the console and with the stand-alone script you have to use `python spectrographic.py [...]` inside the folder containing `spectrographic.py`.

You could also simply import the `SpectroGraphic` class from `spectrographic.base`. Check the doc-strings for detailed explanations and more features.

#### Command-line tool usage
```
usage: spectrographic [-h] [--version] -i PATH_TO_IMAGE [-d DURATION] [-m MIN_FREQ] [-M MAX_FREQ] [-r RESOLUTION] [-c CONTRAST] [-p] [-s SAVE_FILE]

Turn any image into sound.

optional arguments:
  -h, --help            show this help message and exit
  --version             show program's version number and exit
  -i PATH_TO_IMAGE, --image PATH_TO_IMAGE
                        Path of image that we want to embed in a spectrogram.
  -d DURATION, --duration DURATION
                        Duration of generated sound.
  -m MIN_FREQ, --min_freq MIN_FREQ
                        Smallest frequency used for drawing the image.
  -M MAX_FREQ, --max_freq MAX_FREQ
                        Largest frequency used for drawing the image.
  -r RESOLUTION, --resolution RESOLUTION
                        Vertical resolution of the image in the spectrogram.
  -c CONTRAST, --contrast CONTRAST
                        Contrast of the image in the spectrogram.
  -p, --play            Directly play the resulting sound.
  -s SAVE_FILE, --save SAVE_FILE
                        Path to .wav file in which to save the resulting sound.
```
Thus, if you have the source image at `./source.png` and you want to generate a 10s long sound in the frequency range of 10kHz to 20kHz. You also want to save the resulting .wav-file as `sound.wav` and also play the resulting sound. Then you need to run:

`spectrographic --image ./source.png --min_freq 10000 --max_freq 20000 --duration 10 --save sound.wav --play`

or if you are using the stand-alone script:

`python spectrographic.py --image ./source.png --min_freq 10000 --max_freq 20000 --duration 10 --save sound.wav --play`

#### Contribute

Bug reports, fixes and additional features are always welcome! Make sure to run the tests with `python setup.py test` and write your own for new features. Thanks.


CHANGELOG.rst
=========
Changelog
=========

Version 0.8.0
===========

- The core implementation is done
	- base.py contains a working implementation
	of the spetrographic idea.
- Created some examples.
- Adding requirements

Version 0.8.1
===========

- Wrote README.md
- test upload to testpypi
- create demonstrational video

Version 0.8.5
===========

- quick fix of centering in README.md

Version 0.8.6
===========

- added gitpython as dependency

Version 0.8.7
===========

- removed dependencies from config file and moved to requirements

Version 0.9.0
===========
- Release
- upload to the official pypi and making github repo public.

Version 0.9.3
===========

- Put dependencies into setup.cfg so pip can pull them while
  installing


tests/conftest.py
# -*- coding: utf-8 -*-
"""
    Dummy conftest.py for spectrographic.

    If you don't know what this is for, just leave it empty.
    Read more about conftest.py under:
    https://pytest.org/latest/plugins.html
"""

# import pytest


tests/test_skeleton.py
# -*- coding: utf-8 -*-

import pytest
from spectrographic.skeleton import fib

__author__ = "Levi Borodenko"
__copyright__ = "Levi Borodenko"
__license__ = "mit"


def test_fib():
    assert fib(1) == 1
    assert fib(2) == 1
    assert fib(7) == 13
    with pytest.raises(AssertionError):
        fib(-10)


src/spectrographic/base.py
from pathlib import Path

import numpy as np
import simpleaudio as sa
import wavio
from PIL import Image

__author__ = "Levi Borodenko"
__copyright__ = "Levi Borodenko"
__license__ = "mit"


class SpectroGraphic(object):
    """
    Takes an image file and creates a sound that
    draws that image on a spectrogram.

    [description]
    Arguments:
        path {Path} -- Path to file (e.g.: {"./data/python.png"})

    Keyword Arguments:
        height {int} -- y-resolution in spectrogram (default: {100})
        duration {int} -- duration of sound in seconds (default: {20})
        min_freq {int} -- minimal freq. for image (default: {1000})
        max_freq {int} -- maximal freq. for image (default: {8000})
        sample_rate {int} -- Sample rate (default: {44100})
        num_tones {int} -- Number of tones to used to fill in each pixel
        (default: {3})
        contrast {float} -- Contrast between loud and quiet pixels
        (default: {5})
    """

    def __init__(
        self,
        path: Path,
        height: int = 100,
        duration: int = 20,
        min_freq: int = 1000,
        max_freq: int = 8000,
        sample_rate: int = 44100,
        num_tones: int = 3,
        contrast: float = 5,
        use_black_and_white: bool = False,
    ):

        super(SpectroGraphic, self).__init__()
        self.PATH = Path(path)
        self.image = Image.open(self.PATH)
        self.HEIGHT = height
        self.DURATION = duration
        self.SAMPLE_RATE = sample_rate

        # Width after setting height to self.HEIGHT
        # Preserving aspect-ratio
        self.WIDTH = int(self.image.width * (self.HEIGHT / self.image.height))

        # duration per column
        self.DURATION_COL = self.DURATION / self.WIDTH

        # instance of ColumnToSound that will generate
        # the sounds for each column of the image
        self.col_to_sound = ColumnToSound(
            duration=self.DURATION_COL,
            sample_rate=sample_rate,
            min_freq=min_freq,
            max_freq=max_freq,
            y_resolution=height,
            num_tones=num_tones,
            contrast=contrast,
        )

        # Flag whether we have processed the image yet
        self.is_processed = False

        # if true, then we do not use grey-scale with only 0 and 1
        self.USE_BLACK_AND_WHITE = use_black_and_white

    def _resize(self):
        """[summary]
        We resize the image to be at most self.HEIGHT pixels tall.

        [description]
        The reason is to put a limit on the frequency resolution that
        we would need to draw it in on the spectrograph.
        """

        # resizing image
        self.image = self.image.resize(
            size=(self.WIDTH, self.HEIGHT), resample=Image.LANCZOS
        )

    def _preprocess(self):
        """Resizes the image, converts it to grey-scale
        and returns the columns as a np.ndarray.
        """

        # resize image
        self._resize()

        # convert to gray scale
        self.image = self.image.convert(mode="L")

        # get pixels as array and normalise them to be
        # between 0 and 1
        self.image_array = np.array(self.image) / 255

        # transpose image to get list of columns
        self.columns = np.transpose(self.image_array)

        if self.USE_BLACK_AND_WHITE:
            self.columns[self.columns >= 0.5] = 1
            self.columns[self.columns < 0.5] = 0
            self.columns = 1 - self.columns

    def _process(self):
        """Preprocesses the image then turns the
        columns into sounds and stacks them up to produce
        the resulting sound.
        """
        self._preprocess()

        audio_array = np.hstack(
            [self.col_to_sound.gen_soundwall(col) for col in self.columns]
        )

        # convert to 16-bit data
        audio_array *= 32767 / np.max(np.abs(audio_array))
        audio_array = audio_array.astype(np.int16)

        return audio_array

    @property
    def sound_array(self):
        if self.is_processed:
            return self._sound_array
        else:
            self.is_processed = True
            self._sound_array = self._process()
            return self._sound_array

    def play(self):
        """Plays the SpectroGraphic sound.
        """

        # get sound array
        audio = self.sound_array

        # play it using simpleaudio
        wave_object = sa.WaveObject(audio, 1, 2, self.SAMPLE_RATE)
        play_object = wave_object.play()
        play_object.wait_done()

    def save(self, wav_file: Path = "SpectroGraphic.wav"):
        """saves the spectrographic to a .wav file

        We use the wavio module
        """

        wavio.write(wav_file, self.sound_array, self.SAMPLE_RATE)


class ColumnToSound(object):
    """Class to turn grey-scale image columns into
    a sound.

    It takes a numpy array of grey intensities (in the range 0 to 1)
    of length Y_RESOLUTION and turns them into a DURATION seconds long
    sound in the frequency range between MIN_FREQ and MAX_FREQ.


    Arguments:
        duration {int} -- Duration of sound in seconds

    Keyword Arguments:
        sample_rate {int} -- Sample rate of sound (default: {44100})
        min_freq {int} -- Minimal frequency in the spectrograph
        (default: {10000})
        max_freq {int} -- Maximal frequency in the spectrograph
        (default: {17000})
        y_resolution {int} -- Number of pixels to plot (default: {1000})
        num_tones {int} -- Number of tones to use to fill out each pixel
        (default: {3})
        contrast {float} -- Contrast between loud and quiet pixels
        (default: {5})
    """

    def __init__(
        self,
        duration: int,
        sample_rate: int = 44100,
        min_freq: int = 10000,
        max_freq: int = 17000,
        y_resolution: int = 1000,
        num_tones: int = 3,
        contrast: float = 5,
    ):
        super(ColumnToSound, self).__init__()

        # saving imporant parameters
        self.Y_RESOLUTION = y_resolution
        self.CONTRAST = contrast

        # sample rate; 44100 is a good default
        self.SAMPLE_RATE = sample_rate

        # region in which to draw the pixel sound
        self.MIN_FREQ = min_freq
        self.MAX_FREQ = max_freq

        # Number of tones used to fill the pixel sound
        self.NUM_TONES = num_tones

        # frequency window for each pixel sound
        self.HEIGHT = (max_freq - min_freq) / (y_resolution)

        # frequency delta between each tone that fills the pixel sound
        self.tone_delta = self.HEIGHT / num_tones

        # duration in seconds
        self.DURATION = duration

    def _get_wave(self, freq: int, intensity: float = 1, duration: int = 1):
        """Core method that takes a frequency, intensity and duration
        and returns an array representing the corresponding sound.


        Arguments:
            freq {int} -- frequency

        Keyword Arguments:
            intensity {float} -- between 0 and 1 (default: {1})
            duration {float} -- in seconds (default: {1})

        Returns:
            np.ndarray -- sound wave array
        """

        # get timesteps
        t = np.linspace(
            start=0, stop=duration, num=int(duration * self.SAMPLE_RATE), endpoint=False
        )

        # generate corresponding sine wave.
        # this is the only place that CONTRAST acts in
        sound_wave = (intensity ** self.CONTRAST) * np.cos(freq * t * 2 * np.pi)

        return sound_wave

    def pixel_to_sound(self, y: int, intensity: float = 1):
        """Takes a pixel in a imagae column at the y'th position
        from the top and turns it into a sound at a corresponding
        position in the spectrum.

        [description]

        Arguments:
            y {int} -- position of pixel in column from the top.

        Keyword Arguments:
            intensity {float} -- [description] (default: {1})

        Returns:
            np.ndarray -- sound array

        Raises:
            ValueError
        """

        # pixel position (count) from the top must be
        # positive and not larger than the number of pixels
        # in the column.
        if y < 0 or y > self.Y_RESOLUTION:
            raise ValueError("y must be between 0 and 1.")

        # Loudness should be between 0 and 1
        if not (0 <= intensity <= 1):
            raise ValueError("Intensity must be between 0 and 1.")

        # Duration should be positive
        if self.DURATION < 0:
            raise ValueError("Duration must be positive.")

        # calculating base frequency for pixel sound
        base_freq = (self.MAX_FREQ - self.MIN_FREQ) / (self.Y_RESOLUTION) * (
            self.Y_RESOLUTION - y
        ) + self.MIN_FREQ

        # get base wave
        wave = self._get_wave(base_freq, intensity, self.DURATION)

        # add tones to fill up pixel sound
        # first tone:
        tone_freq = base_freq

        # iterating over tones, adding up the sounds.
        for _ in range(self.NUM_TONES):

            tone_freq += self.tone_delta
            wave += self._get_wave(tone_freq, intensity, self.DURATION)

        return wave

    def gen_soundwall(self, column: np.ndarray):
        """Takes a column of pixels and generates
        the sound wall.

        [description]

        Arguments:
            column {np.ndarray} -- Y_RESOLUTION long column of
            pixels (values between 0 and 1)

        Returns:
            np.ndarray -- soundwall
        """

        # empty wave that we will add individual
        # pixel sounds onto
        wave = self.pixel_to_sound(0, 0)

        # iterating over column, adding the pixel sounds
        # of all pixels together to get the final wave.
        for idx, intensity in enumerate(column):
            wave += self.pixel_to_sound(idx, intensity)

        return wave


src/spectrographic/sixel.py
import numpy as np
import simpleaudio as sa


class SoundWall(object):
    """docstring for Sixel"""

    def __init__(
        self,
        sample_rate: int = 44100,
        min_freq: int = 10000,
        max_freq: int = 17000,
        y_resolution: int = 1000,
        num_tones: int = 3,
    ):
        super(SoundWall, self).__init__()

        # Test and save parameters

        # height of image in pixels
        self.y_resolution = y_resolution

        # sample rate; 44100 is a good default
        self.sample_rate = sample_rate

        # region in which to draw the sixel
        self.min_freq = min_freq
        self.max_freq = max_freq

        # Number of tones used to fill the sixel
        self.num_tones = num_tones

        # height of each sixel
        self.height = (max_freq - min_freq) / (y_resolution)

        # delta height between each tone
        self.tone_delta = self.height / num_tones

    def _get_wave(self, freq: int, intensity: float = 1, duration: float = 1):

        # get timesteps, "duration" is note duration in seconds
        t = np.linspace(start=0, stop=duration, num=duration * self.sample_rate)

        sine_wave = intensity * np.sin(freq * t * 2 * np.pi)

        return sine_wave

    def _gen_sixel(self, y: int, intensity: float = 1, duration: float = 1):

        # pixel position (count) from the top
        if y < 0 or y > self.y_resolution:
            raise ValueError("y must be between 0 and 1.")

        # loudness of sixel
        if not (0 <= intensity <= 1):
            raise ValueError("Intensity must be between 0 and 1.")

        # length of sixel
        if duration < 0:
            raise ValueError("Duration must be positive.")

        # calculating base frequency for sixel
        base_freq = (self.max_freq - self.min_freq) / (self.y_resolution) * (
            self.y_resolution - y
        ) + self.min_freq

        # get base wave
        wave = self._get_wave(base_freq, intensity, duration)

        # add tones to fill sixel
        # first tone:
        tone_freq = base_freq
        for _ in range(self.num_tones):

            tone_freq += self.tone_delta
            wave += self._get_wave(tone_freq, intensity, duration)

        return wave


if __name__ == "__main__":

    sw = SoundWall(y_resolution=10, num_tones=10)

    audio = np.hstack(
        [
            sw._gen_sixel(1) + sw._gen_sixel(7),
            sw._gen_sixel(2),
            sw._gen_sixel(3, intensity=0.3, duration=3),
            sw._gen_sixel(4),
            sw._gen_sixel(7),
        ]
    )
    audio *= 32767 / np.max(np.abs(audio))

    # convert to 16-bit data
    audio = audio.astype(np.int16)

    # start playback
    play_obj = sa.play_buffer(audio, 1, 2, sw.sample_rate)

    # wavit for playback to finish before exiting
    play_obj.wait_done()


stand-alone/spectrographic.py
"""[summary]
Stand-alone version of SpectroGraphic, simply
copy this file and use it as the command-line tool.

No pip required.

In the folder with this file, run

python spectrographic.py -h

"""
import argparse
import sys
from pathlib import Path

import numpy as np
import simpleaudio as sa
import wavio
from PIL import Image

__author__ = "Levi Borodenko"
__copyright__ = "Levi Borodenko"
__license__ = "mit"


class SpectroGraphic(object):
    """
    Takes an image file and creates a sound that
    draws that image on a spectrogram.

    [description]
    Arguments:
        path {Path} -- Path to file (e.g.: {"./data/python.png"})

    Keyword Arguments:
        height {int} -- y-resolution in spectrogram (default: {100})
        duration {int} -- duration of sound in seconds (default: {20})
        min_freq {int} -- minimal freq. for image (default: {1000})
        max_freq {int} -- maximal freq. for image (default: {8000})
        sample_rate {int} -- Sample rate (default: {44100})
        num_tones {int} -- Number of tones to used to fill in each pixel
        (default: {3})
        contrast {float} -- Contrast between loud and quiet pixels
        (default: {5})
    """

    def __init__(
        self,
        path: Path,
        height: int = 100,
        duration: int = 20,
        min_freq: int = 1000,
        max_freq: int = 8000,
        sample_rate: int = 44100,
        num_tones: int = 3,
        contrast: float = 5,
        use_black_and_white: bool = False,
    ):

        super(SpectroGraphic, self).__init__()
        self.PATH = Path(path)
        self.image = Image.open(self.PATH)
        self.HEIGHT = height
        self.DURATION = duration
        self.SAMPLE_RATE = sample_rate

        # Width after setting height to self.HEIGHT
        # Preserving aspect-ratio
        self.WIDTH = int(self.image.width * (self.HEIGHT / self.image.height))

        # duration per column
        self.DURATION_COL = self.DURATION / self.WIDTH

        # instance of ColumnToSound that will generate
        # the sounds for each column of the image
        self.col_to_sound = ColumnToSound(
            duration=self.DURATION_COL,
            sample_rate=sample_rate,
            min_freq=min_freq,
            max_freq=max_freq,
            y_resolution=height,
            num_tones=num_tones,
            contrast=contrast,
        )

        # Flag whether we have processed the image yet
        self.is_processed = False

        # if true, then we do not use grey-scale with only 0 and 1
        self.USE_BLACK_AND_WHITE = use_black_and_white

    def _resize(self):
        """[summary]
        We resize the image to be at most self.HEIGHT pixels tall.

        [description]
        The reason is to put a limit on the frequency resolution that
        we would need to draw it in on the spectrograph.
        """

        # resizing image
        self.image = self.image.resize(
            size=(self.WIDTH, self.HEIGHT), resample=Image.LANCZOS
        )

    def _preprocess(self):
        """Resizes the image, converts it to grey-scale
        and returns the columns as a np.ndarray.
        """

        # resize image
        self._resize()

        # convert to gray scale
        self.image = self.image.convert(mode="L")

        # get pixels as array and normalise them to be
        # between 0 and 1
        self.image_array = np.array(self.image) / 255

        # transpose image to get list of columns
        self.columns = np.transpose(self.image_array)

        if self.USE_BLACK_AND_WHITE:
            self.columns[self.columns >= 0.5] = 1
            self.columns[self.columns < 0.5] = 0
            self.columns = 1 - self.columns

    def _process(self):
        """Preprocesses the image then turns the
        columns into sounds and stacks them up to produce
        the resulting sound.
        """
        self._preprocess()

        audio_array = np.hstack(
            [self.col_to_sound.gen_soundwall(col) for col in self.columns]
        )

        # convert to 16-bit data
        audio_array *= 32767 / np.max(np.abs(audio_array))
        audio_array = audio_array.astype(np.int16)

        return audio_array

    @property
    def sound_array(self):
        if self.is_processed:
            return self._sound_array
        else:
            self.is_processed = True
            self._sound_array = self._process()
            return self._sound_array

    def play(self):
        """Plays the SpectroGraphic sound.
        """

        # get sound array
        audio = self.sound_array

        # play it using simpleaudio
        wave_object = sa.WaveObject(audio, 1, 2, self.SAMPLE_RATE)
        play_object = wave_object.play()
        play_object.wait_done()

    def save(self, wav_file: Path = "SpectroGraphic.wav"):
        """saves the spectrographic to a .wav file

        We use the wavio module
        """

        wavio.write(wav_file, self.sound_array, self.SAMPLE_RATE)


class ColumnToSound(object):
    """Class to turn grey-scale image columns into
    a sound.

    It takes a numpy array of grey intensities (in the range 0 to 1)
    of length Y_RESOLUTION and turns them into a DURATION seconds long
    sound in the frequency range between MIN_FREQ and MAX_FREQ.


    Arguments:
        duration {int} -- Duration of sound in seconds

    Keyword Arguments:
        sample_rate {int} -- Sample rate of sound (default: {44100})
        min_freq {int} -- Minimal frequency in the spectrograph
        (default: {10000})
        max_freq {int} -- Maximal frequency in the spectrograph
        (default: {17000})
        y_resolution {int} -- Number of pixels to plot (default: {1000})
        num_tones {int} -- Number of tones to use to fill out each pixel
        (default: {3})
        contrast {float} -- Contrast between loud and quiet pixels
        (default: {5})
    """

    def __init__(
        self,
        duration: int,
        sample_rate: int = 44100,
        min_freq: int = 10000,
        max_freq: int = 17000,
        y_resolution: int = 1000,
        num_tones: int = 3,
        contrast: float = 5,
    ):
        super(ColumnToSound, self).__init__()

        # saving imporant parameters
        self.Y_RESOLUTION = y_resolution
        self.CONTRAST = contrast

        # sample rate; 44100 is a good default
        self.SAMPLE_RATE = sample_rate

        # region in which to draw the pixel sound
        self.MIN_FREQ = min_freq
        self.MAX_FREQ = max_freq

        # Number of tones used to fill the pixel sound
        self.NUM_TONES = num_tones

        # frequency window for each pixel sound
        self.HEIGHT = (max_freq - min_freq) / (y_resolution)

        # frequency delta between each tone that fills the pixel sound
        self.tone_delta = self.HEIGHT / num_tones

        # duration in seconds
        self.DURATION = duration

    def _get_wave(self, freq: int, intensity: float = 1, duration: float = 1):
        """Core method that takes a frequency, intensity and duration
        and returns an array representing the corresponding sound.


        Arguments:
            freq {int} -- frequency

        Keyword Arguments:
            intensity {float} -- between 0 and 1 (default: {1})
            duration {float} -- in seconds (default: {1})

        Returns:
            np.ndarray -- sound wave array
        """

        # get timesteps
        t = np.linspace(
            start=0, stop=duration, num=int(duration * self.SAMPLE_RATE), endpoint=False
        )

        # generate corresponding sine wave.
        # this is the only place that CONTRAST acts in
        sound_wave = (intensity ** self.CONTRAST) * np.cos(freq * t * 2 * np.pi)

        return sound_wave

    def pixel_to_sound(self, y: int, intensity: float = 1):
        """Takes a pixel in a imagae column at the y'th position
        from the top and turns it into a sound at a corresponding
        position in the spectrum.

        [description]

        Arguments:
            y {int} -- position of pixel in column from the top.

        Keyword Arguments:
            intensity {float} -- [description] (default: {1})

        Returns:
            np.ndarray -- sound array

        Raises:
            ValueError
        """

        # pixel position (count) from the top must be
        # positive and not larger than the number of pixels
        # in the column.
        if y < 0 or y > self.Y_RESOLUTION:
            raise ValueError("y must be between 0 and 1.")

        # Loudness should be between 0 and 1
        if not (0 <= intensity <= 1):
            raise ValueError("Intensity must be between 0 and 1.")

        # Duration should be positive
        if self.DURATION < 0:
            raise ValueError("Duration must be positive.")

        # calculating base frequency for pixel sound
        base_freq = (self.MAX_FREQ - self.MIN_FREQ) / (self.Y_RESOLUTION) * (
            self.Y_RESOLUTION - y
        ) + self.MIN_FREQ

        # get base wave
        wave = self._get_wave(base_freq, intensity, self.DURATION)

        # add tones to fill up pixel sound
        # first tone:
        tone_freq = base_freq

        # iterating over tones, adding up the sounds.
        for _ in range(self.NUM_TONES):

            tone_freq += self.tone_delta
            wave += self._get_wave(tone_freq, intensity, self.DURATION)

        return wave

    def gen_soundwall(self, column: np.ndarray):
        """Takes a column of pixels and generates
        the sound wall.

        [description]

        Arguments:
            column {np.ndarray} -- Y_RESOLUTION long column of
            pixels (values between 0 and 1)

        Returns:
            np.ndarray -- soundwall
        """

        # empty wave that we will add individual
        # pixel sounds onto
        wave = self.pixel_to_sound(0, 0)

        # iterating over column, adding the pixel sounds
        # of all pixels together to get the final wave.
        for idx, intensity in enumerate(column):
            wave += self.pixel_to_sound(idx, intensity)

        return wave


def parse_args(args):
    """Parse command line parameters

    Args:
      args ([str]): command line parameters as list of strings

    Returns:
      :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = argparse.ArgumentParser(description="Turn any image into sound.")

    parser.add_argument(
        "-i",
        "--image",
        dest="path_to_image",
        help="Path to image",
        type=Path,
        action="store",
        required=True,
    )
    parser.add_argument(
        "-d",
        "--duration",
        dest="duration",
        help="Duration of generated sound",
        action="store",
        default=20,
        type=int,
    )
    parser.add_argument(
        "-m",
        "--min_freq",
        dest="min_freq",
        help="Minimal frequency used in SpectroGraphic",
        action="store",
        default=500,
        type=int,
    )
    parser.add_argument(
        "-M",
        "--max_freq",
        dest="max_freq",
        help="Maximal frequency used in SpectroGraphic",
        action="store",
        default=7500,
        type=int,
    )
    parser.add_argument(
        "-r",
        "--resolution",
        dest="resolution",
        help="y-resolution of SpectroGraphic",
        action="store",
        default=150,
        type=int,
    )
    parser.add_argument(
        "-c",
        "--contrast",
        dest="contrast",
        help="Contrast of SpectroGraphic",
        action="store",
        default=3,
        type=int,
    )
    parser.add_argument(
        "-p",
        "--play",
        action="store_true",
        dest="play",
        help="Play the Spectrographic sound",
    )
    parser.add_argument(
        "-s",
        "--save",
        dest="save_file",
        help="Path to .wav file in which to save the SpectroGraphic.",
        action="store",
        default=Path("SoundGraphic.wav"),
        type=Path,
    )
    return parser.parse_args(args)


def main(args):
    """Main entry point allowing external calls

    Args:
      args ([str]): command line parameter list
    """
    args = parse_args(args)

    sg = SpectroGraphic(
        path=args.path_to_image,
        height=args.resolution,
        duration=args.duration,
        min_freq=args.min_freq,
        max_freq=args.max_freq,
        contrast=args.contrast,
    )

    if args.play:
        sg.play()

    sg.save(wav_file=args.save_file)


def run():
    """Entry point for console_scripts
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    run()


src/spectrographic/cli.py
# -*- coding: utf-8 -*-
"""
This is a skeleton file that can serve as a starting point for a Python
console script. To run this script uncomment the following lines in the
[options.entry_points] section in setup.cfg:

    console_scripts =
         spectrographic = spectrographic.cli:run

Then run `python setup.py install` which will install the command `spectrographic`
inside your current environment.
"""

import argparse
import sys
from pathlib import Path

from spectrographic import __version__
from spectrographic.base import SpectroGraphic

__author__ = "Levi Borodenko"
__copyright__ = "Levi Borodenko"
__license__ = "mit"


def parse_args(args):
    """Parse command line parameters

    Args:
      args ([str]): command line parameters as list of strings

    Returns:
      :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = argparse.ArgumentParser(
        description="Turn any image into sound.", epilog="By Levi B."
    )
    parser.add_argument(
        "--version",
        action="version",
        version="spectrographic {ver}".format(ver=__version__),
    )
    parser.add_argument(
        "-i",
        "--image",
        dest="path_to_image",
        help="Path of image that we want to embed in a spectrogram.",
        type=Path,
        action="store",
        required=True,
    )
    parser.add_argument(
        "-d",
        "--duration",
        dest="duration",
        help="Duration of generated sound.",
        action="store",
        default=20,
        type=int,
    )
    parser.add_argument(
        "-m",
        "--min_freq",
        dest="min_freq",
        help="Smallest frequency used for drawing the image.",
        action="store",
        default=500,
        type=int,
    )
    parser.add_argument(
        "-M",
        "--max_freq",
        dest="max_freq",
        help="Largest frequency used for drawing the image.",
        action="store",
        default=7500,
        type=int,
    )
    parser.add_argument(
        "-r",
        "--resolution",
        dest="resolution",
        help="Vertical resolution of the image in the spectrogram.",
        action="store",
        default=150,
        type=int,
    )
    parser.add_argument(
        "-c",
        "--contrast",
        dest="contrast",
        help="Contrast of the image in the spectrogram.",
        action="store",
        default=3,
        type=int,
    )
    parser.add_argument(
        "-p",
        "--play",
        action="store_true",
        dest="play",
        help="Directly play the resulting sound.",
    )
    parser.add_argument(
        "-s",
        "--save",
        dest="save_file",
        help="Path to .wav file in which to save the resulting sound.",
        action="store",
        default="SoundGraphic.wav",
        type=str,
    )
    return parser.parse_args(args)


def main(args):
    """Main entry point allowing external calls

    Args:
      args ([str]): command line parameter list
    """
    args = parse_args(args)

    sg = SpectroGraphic(
        path=args.path_to_image,
        height=args.resolution,
        duration=args.duration,
        min_freq=args.min_freq,
        max_freq=args.max_freq,
        contrast=args.contrast,
    )

    if args.play:
        sg.play()

    sg.save(wav_file=args.save_file)


def run():
    """Entry point for console_scripts
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    run()
